nixos/modules/services/home-automation/home-assistant.nix

```diff
@@ -42,33 +35,72 @@ let
   #   platform = "mqtt";
   #   ...
   # } ];
-  useComponentPlatform = component: elem component (usedPlatforms cfg.config);
+  usedPlatforms = config:
+    if isAttrs config then
+      optional (config ? platform) config.platform
+      ++ concatMap usedPlatforms (attrValues config)
+    else if isList config then
+      concatMap usedPlatforms config
+    else [ ];
 
-  useExplicitComponent = component: elem component explicitComponents;
+  useComponentPlatform = component: elem component (usedPlatforms cfg.config);
 
-  # Returns whether component is used in config or explicitly passed into package
+  # Returns whether component is used in config, explicitly passed into package or
+  # configured in the module.
   useComponent = component:
     hasAttrByPath (splitString "." component) cfg.config
     || useComponentPlatform component
-    || useExplicitComponent component;
+    || useExplicitComponent component
+    || builtins.elem component cfg.extraComponents;
 
-  # List of components used in config
+  # Final list of components passed into the package to include required dependencies
   extraComponents = filter useComponent availableComponents;
+  extraComponentsFile = format.generate "extraComponents.yaml" extraComponents;
 
-  package = if (cfg.autoExtraComponents && cfg.config != null)
-    then (cfg.package.override { inherit extraComponents; })
+  # Final list of extra packages passed into the package, that also includes custom
+  # component dependencies
+  extraPackages = cfg.extraPackages
+    ++ map (component: component.dependencies or []) cfg.customComponents;
```

Comment:
We somehow need to propagate python dependencies that custom components have into the final home-assistant package. The silly idea is to define `passthru.dependencies`, wondering if there is a more established way of handling these kinds of relationships in nixpkgs.